### ISSUE TITLE
synthetic shape + magic-point train question.

### ISSUE THREAD

**Issue by pervin0527** [rpautrat/SuperPoint#265]
hi, I'm referencing your repository and rewriting the code with tf2 now.
when i train MagicPoint with Synthetic Shapes, the train result are not bad.
However, as you can see below, trained MagicPoint detecting keypoints on some Ellipse.
![0015](https://user-images.githubusercontent.com/43873288/186629063-db17f641-3468-422d-9e5d-4d402a9cf6bb.png)
![0058](https://user-images.githubusercontent.com/43873288/186629120-446fff47-0e56-4cd3-905d-d8eadcd8c35b.png)

this is training steps.
- img_size : [240, 320]
- nms : 4
- threshold : 0.1
- parameters are same as your's. can see [here](https://github.com/pervin0527/pervinco/blob/master/source/SuperPoint/configs/magic-point_train.yaml)
![Screenshot from 2022-08-25 18-31-14](https://user-images.githubusercontent.com/43873288/186629699-e1fb5c79-b27b-4b6f-931a-5f64f374dd83.png)

i guess the reason that sparse categorical cross entropy is changed in tf2, so it can't apply Valid-Mask to weight.
- [tf1 sparse-CE](https://www.tensorflow.org/api_docs/python/tf/compat/v1/losses/sparse_softmax_cross_entropy)
- [tf2 sparse-CE](https://www.tensorflow.org/api_docs/python/tf/nn/sparse_softmax_cross_entropy_with_logits)
can you give me some advices to improve magicpoint's performance???

this is my repo's link.
[https://github.com/pervin0527/pervinco/tree/master/source/SuperPoint](https://github.com/pervin0527/pervinco/tree/master/source/SuperPoint)

**Comment by rpautrat** [comment]
Hi, I never used TF2, so can't help on specific details regarding it. But it should be fairly easy to integrate a valid mask manually. When computing the loss, you can compute the cross entropy for all pixels, and then mask out the pixels that are outside of the mask and ignore them in the loss.

**Comment by pervin0527** [comment]
> Hi, I never used TF2, so can't help on specific details regarding it. But it should be fairly easy to integrate a valid mask manually. When computing the loss, you can compute the cross entropy for all pixels, and then mask out the pixels that are outside of the mask and ignore them in the loss.

@rpautrat 
hmm...I'm sorry, I can't understand what you mean.(I can't speak English very well. so sorry...)

def detector_loss(keypoint_map, logits, valid_mask):
    labels = tf.cast(keypoint_map[..., tf.newaxis], tf.float32) 
    labels = tf.nn.space_to_depth(labels, 8)

    shape = tf.concat([tf.shape(labels)[:3], [1]], axis=0)
    labels = tf.concat([2*labels, tf.ones(shape)], 3)
    labels = tf.argmax(labels + tf.random.uniform(tf.shape(labels), 0, 0.1), axis=3)

    valid_mask = tf.ones_like(keypoint_map) if valid_mask is None else valid_mask
    valid_mask = tf.cast(valid_mask[..., tf.newaxis], tf.float32)
    valid_mask = tf.nn.space_to_depth(valid_mask, 8)
    valid_mask = tf.reduce_prod(valid_mask, axis=3)

    loss = tf.nn.sparse_softmax_cross_entropy_with_logits(logits=logits, labels=labels)

this is the loss function part. this is exactly same your code except last line.
I tried the following two methods, is this correct??
try 1. loss = tf.math.divide_no_nan(tf.reduce_sum(loss * valid_mask), tf.reduce_sum(valid_mask))
try 2. loss *= valid_mask

**Comment by rpautrat** [comment]
Yes, I think your first try should be correct.

**Comment by pervin0527** [comment]
> 

Thank you! i will try one more time with first method!!
